### PR TITLE
JointDistribution & UniformOrderStatistics

### PR DESCRIPTION
**Pull request by regislebrun**
In this PR, you will find:
  * An extension of the JointDistribution class. The decomposition 1D margins/copula is not the only way to define joint distributions using composition. The copula can be replaced by any other nD distribution defined over $[0,1]^n$, the resulting composition will be a valid joint distribution. The only point is that the 1D distributions will no more be the 1D margins of the joint distribution. Such an extension is mandatory e.g to build the joint distribution of general order statistics from the joint distribution of uniform order statistics.
  * A new distribution called the UniformOrderStatistics, which is the joint distribution of the order statistics of $n$ iid random variables uniformly distributed over $[0, 1]$.
  * A bug fix in Mixture::getParameterCollection()
  * A bug fix in Dirichlet::getSkewness()
  * A bug fix in DistributionImplementation::getKendallTau(), with a repercussion on many classes

### CODE REVIEW

File: lib/src/Uncertainty/Distribution/openturns/JointDistribution.hxx [hunk 1]
```diff
@@ -71,12 +71,15 @@ public:
   void setDistributionCollection(const DistributionCollection & coll);
   DistributionCollection getDistributionCollection() const;
 
+  /** Kernel accessor */
+  void setKernel(const Distribution & kernel);
+  Distribution getKernel() const;
+
   /** Copula accessor */
   void setCopula(const Distribution & copula);
   Distribution getCopula() const override;
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/openturns/JointDistribution.hxx:80]: so setCopula should be deprecated ?

Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/openturns/JointDistribution.hxx:80]: I don't think so. A copula is a valid kernel and is more explicit when one want to explicitely use this class to build a joint distribution based on Sklar's theorem.
Another option is to use this class as a base class for others, one based on copulas, one based on uniform order statistics and so one.

File: python/src/UniformOrderStatistics_doc.i.in [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+>>> import openturns.experimental as otexp
+>>> distribution = otexp.UniformOrderStatistics(3)
+
+Draw a sample:
+
+>>> sample = distribution.getSample(5)
+
+Compute some quantities:
+
+>>> pdf = distribution.computePDF([0.1, 0.5])
```
Comment (jschueller) [inline python/src/UniformOrderStatistics_doc.i.in:49]: seems the dimension does not match

File: python/src/UniformOrderStatistics_doc.i.in [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+This is the joint distribution of the order statistics :math:`(X_{(1)},\dots,X_{(n)})` of a sample :math:`(X_1,\dots, X_n)` of :math:`n` independent random variables uniformly distributed over :math:`[0,1]`. Its probability density function is defined as:
+
+.. math::
+
+    f_X(x) = n!\,\mathbf{1}_{\cS}(x)
+
+where :math:`\cS\subset\Rset^n` is the simplex with vertices :math:`\cV`:
+
+.. math::
+   \cV=\left\{(\underbrace{0,\dots,0}_{k}, \underbrace{1,\dots,1}_{n-k}), k\in\{0,\dots,n\}\right\}`.
```
Comment (mbaudin47) [inline python/src/UniformOrderStatistics_doc.i.in:28]:
```suggestion
   \cV=\left\{(\underbrace{0,\dots,0}_{k}, \underbrace{1,\dots,1}_{n-k}), \; k\in\{0,\dots,n\}\right\}`.
```

File: python/doc/pyplots/UniformOrderStatistics.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import openturns as ot
+import openturns.experimental as otexp
+from openturns.viewer import View
+
+grid = ot.GridLayout(1, 2)
+pdf_2d = ot.Graph("Uniform order statistices, PDF", "x1", "x2", True)
```
Comment (mbaudin47) [inline python/doc/pyplots/UniformOrderStatistics.py:6]:
```suggestion
pdf_2d = ot.Graph("Uniform order statistics, PDF", "x1", "x2", True)
```

File: python/src/UniformOrderStatistics_doc.i.in [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+dimension : int
+    The dimension of the distribution.
+
+See also
+--------
+openturns.MaximumEntropyOrderStatisticsDistribution
+
+Notes
+-----
+This is the joint distribution of the order statistics :math:`(X_{(1)},\dots,X_{(n)})` of a sample :math:`(X_1,\dots, X_n)` of :math:`n` independent random variables uniformly distributed over :math:`[0,1]`. Its probability density function is defined as:
```
Comment (mbaudin47) [inline python/src/UniformOrderStatistics_doc.i.in:19]: Splitting the text onto several lines will make the code easier to read.
Moreover, a small mathematical text will help guide users through the mathematical meaning of the class.
```suggestion
This is the joint distribution of the order statistics :math:`(X_{(1)},\dots,X_{(n)})`
of a sample :math:`(X_1,\dots, X_n)` of :math:`n` independent random variables
uniformly distributed over :math:`[0,1]`.
In other words:

- :math:`X_i \sim \mathcal{U}(0,1)` for :math:`i \in \set{1, ..., \inpuDim}` ;
- :math:`X_{(1)} \leq X_{(2)} \leq \ldots \leq X_{(\inputDim)}`.

Its probability density function is defined as:
```

File: python/src/UniformOrderStatistics_doc.i.in [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+>>> import openturns.experimental as otexp
+>>> distribution = otexp.UniformOrderStatistics(3)
+
+Draw a sample:
+
+>>> sample = distribution.getSample(5)
+
+Compute some quantities:
+
+>>> pdf = distribution.computePDF([0.1, 0.15, 0.45])
```
Comment (mbaudin47) [inline python/src/UniformOrderStatistics_doc.i.in:49]:
```suggestion
>>> pdf = distribution.computePDF([0.1, 0.15, 0.45])  // Notice the order
```

File: python/src/CMakeLists.txt [hunk 7]
```diff
@@ -1024,11 +1024,12 @@ ot_add_python_module(experimental experimental_module.i
                      FieldFunctionalChaosSobolIndices.i FieldFunctionalChaosSobolIndices_doc.i.in
                      UserDefinedMetropolisHastings.i UserDefinedMetropolisHastings_doc.i.in
                      SmolyakExperiment.i SmolyakExperiment_doc.i.in
-		     SmoothedUniformFactory.i SmoothedUniformFactory_doc.i.in
+		             SmoothedUniformFactory.i SmoothedUniformFactory_doc.i.in
```
Comment (mbaudin47) [inline python/src/CMakeLists.txt:1027]: Is that necessary?

File: python/doc/pyplots/UniformOrderStatistics.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+cdf_2d = ot.Graph("Uniform order statistics, CDF", "x1", "x2", True)
+
+distribution_2d = otexp.UniformOrderStatistics(2)
+
+cloud = ot.Cloud(distribution_2d.getSample(1000))
+cloud.setColor("blue")
```
Comment (mbaudin47) [inline python/doc/pyplots/UniformOrderStatistics.py:12]: With the help of #2422, this is not necessary anymore.
```suggestion
```

File: python/src/JointDistribution_doc.i.in [hunk 9]
```diff
@@ -6,22 +6,24 @@ Parameters
...
 Notes
 -----
 A JointDistribution is a :math:`n`-dimensional distribution which can be
-written in terms of 1-d marginal distribution functions and a copula :math:`C`
-which describes the dependence structure between the variables.
-Its cumulative distribution function :math:`F` is defined by its marginal
-distributions :math:`F_i` and the copula :math:`C` through the relation:
+written in terms of 1-d distribution functions and a kernel :math:`K`
+which describes the interaction between the variables. If :math:`K` is a
+copula, then it represents exactly the dependence structure of the joint
```
Comment (mbaudin47) [inline python/src/JointDistribution_doc.i.in:19]: In mathematics, an object represents or does not represents something. It never _approximately_ represents something.
```suggestion
copula, then it represents the dependence structure of the joint
```

File: python/src/JointDistribution_doc.i.in [hunk 10]
```diff
@@ -6,22 +6,24 @@ Parameters
 distributions : list of :class:`~openturns.Distribution`
     List of :math:`n` marginals of the distribution. Each marginal must be of
     dimension 1.
-copula : :class:`~openturns.Distribution`, optional
-    A copula. If not mentioned, the copula is set to an
+kernel : :class:`~openturns.Distribution`, optional
+    A distribution over :math:`[0,1]^n`. If not mentioned, the kernel is set to an
```
Comment (mbaudin47) [inline python/src/JointDistribution_doc.i.in:10]:
```suggestion
    A distribution over :math:`[0,1]^\inputDim`. If not mentioned, the kernel is set to an
```

File: python/src/JointDistribution_doc.i.in [hunk 11]
```diff
@@ -6,22 +6,24 @@ Parameters
...
+which describes the interaction between the variables. If :math:`K` is a
+copula, then it represents exactly the dependence structure of the joint
+distribution and the given 1-d distributions are its marginal distributions.
+Its cumulative distribution function :math:`F` is defined by the 1-d
+distributions :math:`F_i` and the kernel :math:`K` through the relation:
 
 .. math::
 
-    F(x_1, \cdots, x_n) = C(F_1(x_1), \cdots, F_n(x_n))
+    F(x_1, \cdots, x_n) = K(F_1(x_1), \cdots, F_n(x_n))
```
Comment (mbaudin47) [inline python/src/JointDistribution_doc.i.in:26]:
```suggestion
    F(x_1, \cdots, x_\inputDim) = K(F_1(x_1), \cdots, F_n(x_\inputDim))
```

File: lib/src/Uncertainty/Distribution/UniformOrderStatistics.cxx [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+/* Get the quantile of the distribution */
+Scalar UniformOrderStatistics::computeScalarQuantile(const Scalar prob,
+    const Bool tail) const
+{
+  if (tail) return 1.0 - prob;
+  return prob;
+}
+
+/* Get the i-th marginal distribution */
+Distribution UniformOrderStatistics::getMarginal(const UnsignedInteger i) const
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Distribution/UniformOrderStatistics.cxx:379]: According to (Arnold et al., 2008) eq. 2.2.15 page 13, this is a Beta(i, n - i + 1), isn'it?

File: lib/src/Uncertainty/Distribution/UniformOrderStatistics.cxx [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+  return getMarginal(Indices(1, i));
+}
+
+/* Get the distribution of the marginal distribution corresponding to indices dimensions */
+Distribution UniformOrderStatistics::getMarginal(const Indices & indices) const
+{
+  const UnsignedInteger dimension = getDimension();
+  if (!indices.check(dimension)) throw InvalidArgumentException(HERE) << "The indices of a marginal distribution must be in the range [0, dim-1] and must be different";
+  if (dimension == 1) return clone();
+  if (indices.getSize() == 1)
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Distribution/UniformOrderStatistics.cxx:391]: Shouldn't this code be based on the previous `getMarginal(UnsignedInteger i)`?

File: lib/src/Uncertainty/Distribution/UniformOrderStatistics.cxx [hunk 14]
```diff
@@ -0,0 +1,470 @@
...
+  {
+    const UnsignedInteger i = indices[0];
+    Beta marginal(i + 1.0, dimension_ - i, 0.0, 1.0);
+    marginal.setDescription(Description(1, getDescription()[i]));
+    return marginal;
+  }
+  return DistributionImplementation::getMarginal(indices);
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Distribution/UniformOrderStatistics.cxx:398]: According to (Arnold et al., 2008) pages 19 (eq. 2.3.12) and page 78, in the case where d = 2, then the distribution is Bivariate Beta(i, j - i, n - j = 1). Is this distribution in the catalogue?

**Review by mbaudin47 — CHANGES_REQUESTED** [review]
Thank you very much for this implementation. A unit test of UniformOrderStatistics will make this code even more convincing.

File: lib/src/Uncertainty/Distribution/UniformOrderStatistics.cxx (hunk 12 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/UniformOrderStatistics.cxx:379]: No. You have to take into account the shift of 1 between the C++/Python indexing and the textbook indexing.

File: lib/src/Uncertainty/Distribution/UniformOrderStatistics.cxx (hunk 14 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/UniformOrderStatistics.cxx:398]: We don't have this in stock! The  closest is the bivariate Dirichlet distribution, but there is a transformation to be applied to the arguments of the PDF to have the correct support.
BTW I am waiting for https://github.com/openturns/openturns/pull/2603 to be merged in order to implement a dedicated distribution for these multivariate marginal distributions. We already detailed the formulas for the PDF and CDF with @adutfoy and we need the new version of MarginalDistribution to have a base for these new distributions.

File: lib/src/Uncertainty/Distribution/openturns/UniformOrderStatistics.hxx [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+  : public ContinuousDistribution
+{
+  CLASSNAME
+public:
+
+  /** Default constructor */
+  UniformOrderStatistics();
+
+  /** Parameters constructor */
+  UniformOrderStatistics(const UnsignedInteger dimension);
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/openturns/UniformOrderStatistics.hxx:44]: explicit

Comment (jschueller) [reply lib/src/Uncertainty/Distribution/openturns/UniformOrderStatistics.hxx:44]: you could even set a default value so that the default ctor can be removed

File: lib/src/Uncertainty/Distribution/openturns/JointDistribution.hxx [hunk 16]
```diff
@@ -71,12 +71,15 @@ public:
   void setDistributionCollection(const DistributionCollection & coll);
   DistributionCollection getDistributionCollection() const;
 
+  /** Kernel accessor */
+  void setKernel(const Distribution & kernel);
+  Distribution getKernel() const;
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/openturns/JointDistribution.hxx:76]: the doc is missing

File: lib/src/Uncertainty/Distribution/UniformOrderStatistics.cxx (hunk 12 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/UniformOrderStatistics.cxx:379]: Sorry, my comment is not relevant. I did not carefully read the `getMarginal(Indices(1, i))`: this triggers the code below which takes into account this particular case.

File: python/src/UniformOrderStatistics_doc.i.in [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+- let :math:`\sampleSize \in \Nset` be the sample size ;
+- let :math:`X_i \sim \mathcal{U}(0,1)` for :math:`i \in \set{1, ..., \sampleSize}` be independent random variables ;
+- order them in increasing order: :math:`X_{(1)} \leq X_{(2)} \leq \ldots \leq X_{(\sampleSize)}`.
+
+Its probability density function is defined as (see [arnold2008]_ eq. 2.2.4 page 11):
+
+.. math::
+
+    f_X(x) = n!\,\mathbf{1}_{\cS}(x)
```
Comment (l-brevault) [inline python/src/UniformOrderStatistics_doc.i.in:32]: The notation f_X(x) might be confusing with the classical notation of the PDF of X, should we introduce another notation for the PDF of the order statistics ?

"Its probability density function (noted ......) is defined as (see [arnold2008]_ eq. 2.2.4 page 11):"

File: python/src/UniformOrderStatistics_doc.i.in [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+This is the joint distribution of the order statistics :math:`(X_{(1)},\dots,X_{(n)})`
+of a sample :math:`(X_1,\dots, X_n)` of :math:`n` independent random variables
+uniformly distributed over :math:`[0,1]`.
+In other words:
+
+- let :math:`\sampleSize \in \Nset` be the sample size ;
+- let :math:`X_i \sim \mathcal{U}(0,1)` for :math:`i \in \set{1, ..., \sampleSize}` be independent random variables ;
+- order them in increasing order: :math:`X_{(1)} \leq X_{(2)} \leq \ldots \leq X_{(\sampleSize)}`.
+
+Its probability density function is defined as (see [arnold2008]_ eq. 2.2.4 page 11):
```
Comment (l-brevault) [inline python/src/UniformOrderStatistics_doc.i.in:28]: [arnold2008] should it be added in bibliography.rst ?

**Review by l-brevault — APPROVED** [review]
Thank you for this very nice addition ! (just two minor remarks on the doc)

File: python/src/UniformOrderStatistics_doc.i.in (hunk 17 above)
Comment (regislebrun) [reply python/src/UniformOrderStatistics_doc.i.in:23]: @l-brevault Thanks for the reference. I changed the notation of the random vector of order statistics using something github is not able to render...

**Review by sofianehaddad — APPROVED** [review]

**Comment by regislebrun** [comment]
Some significant bugs have been found in Dirichlet, Mixture, DistributionImplementation::getKendallTau() since the last review, explaining the delay.

File: lib/src/Uncertainty/Distribution/JointDistribution.cxx [hunk 19]
```diff
@@ -81,14 +81,14 @@ JointDistribution::JointDistribution(const DistributionCollection & coll)
...
+                                     const Distribution & kernel)
   : DistributionImplementation()
   , distributionCollection_()
-  , copula_(copula)
+  , kernel_(kernel)
 {
   setName("JointDistribution");
-  if (!copula.isCopula()) throw InvalidArgumentException(HERE) << "Error: the given distribution=" << copula << " is not a copula.";
-  setDimension(copula.getDimension());
+  if (!(kernel.getRange() == Interval(kernel.getDimension()))) throw InvalidArgumentException(HERE) << "Error: the given kernel=" << kernel << " has a range=" << kernel.getRange() << " different from the unit hypercube.";
```
Comment (josephmure) [inline lib/src/Uncertainty/Distribution/JointDistribution.cxx:90]: Perhaps it is too late for that but I would propose the name "pseudo-copula" or "semi-copula" instead of "kernel". IMO it is better than the (too) generic "kernel" because it tells that this object shares some properties with a copula (ie is a multi-dimensional distribution on the unit cube).

File: lib/src/Uncertainty/Distribution/JointDistribution.cxx [hunk 20]
```diff
@@ -308,6 +323,32 @@ JointDistribution::DistributionCollection JointDistribution::getDistributionColl
...
+      throw InvalidArgumentException(HERE) << "Kernel has a dimension different from the JointDistribution's";
+  }
+  else
+    setDimension(kernel.getDimension());
+
+  kernel_ = kernel;
+  isAlreadyComputedCovariance_ = false;
+  // We ensure that the kernel has the same description than the JointDistribution
+  kernel_.setDescription(getDescription());
+}
```
Comment (josephmure) [inline lib/src/Uncertainty/Distribution/JointDistribution.cxx:343]: Couldn't you check that the support of the "kernel" (see my post above for other name ideas) is the unit cube?

File: lib/src/Uncertainty/Distribution/JointDistribution.cxx [hunk 21]
```diff
@@ -344,17 +387,17 @@ JointDistribution * JointDistribution::clone() const
...
 {
   const UnsignedInteger dimension = getDimension();
+  // Complex case: use 1D CDF inversion
+  if (!kernel_.isCopula() || !hasIndependentCopula())
+  {
+    Point realization(kernel_.getRealization());
+    for (UnsignedInteger i = 0; i < dimension; ++i) realization[i] = distributionCollection_[i].computeScalarQuantile(realization[i]);
+    return realization;
+  } // General case
   if (dimension == 1) return distributionCollection_[0].getRealization();
```
Comment (josephmure) [inline lib/src/Uncertainty/Distribution/JointDistribution.cxx:397]: Shouldn't this case be handled at the start of getRealization, since it is the trivial case?

File: lib/src/Uncertainty/Distribution/JointDistribution.cxx [hunk 22]
```diff
@@ -344,17 +387,17 @@ JointDistribution * JointDistribution::clone() const
...
   Point result(dimension);
-  if (hasIndependentCopula())
-  {
-    for (UnsignedInteger i = 0; i < dimension; ++i) result[i] = distributionCollection_[i].getRealization()[0];
-    return result;
-  }
-  // General case
-  const Point realization(copula_.getRealization());
-  for (UnsignedInteger i = 0; i < dimension; ++i) result[i] = distributionCollection_[i].computeScalarQuantile(realization[i]);
+  for (UnsignedInteger i = 0; i < dimension; ++i) result[i] = distributionCollection_[i].getRealization()[0];
```
Comment (josephmure) [inline lib/src/Uncertainty/Distribution/JointDistribution.cxx:400]: I am surprised by this bit of code. Maybe I missed it, but I don't see where you handle the case where `kernel_`*is* a copula. I saw:

- the complex case where the `kernel_`is not a copula *and* the copula is not independent
- the trivial 1d case
- the independent copula case

Where do you handle the cases where:
- `kernel_`is not a copula but has an independent copula
- `kernel_`is a copula but is not independent

File: lib/src/Uncertainty/Distribution/JointDistribution.cxx [hunk 23]
```diff
@@ -448,95 +491,96 @@ Point JointDistribution::computeDDF(const Point & point) const
...
   if (point.getDimension() != dimension) throw InvalidArgumentException(HERE) << "Error: the given point must have dimension=" << dimension << ", here dimension=" << point.getDimension();
 
+  if (!kernel_.isCopula() || !hasIndependentCopula())
+  {
+    // General case
+    Point uPoint(dimension);
+    for (UnsignedInteger i = 0; i < dimension; ++i) uPoint[i] = distributionCollection_[i].computeCDF(point[i]);
+    return kernel_.computeCDF(uPoint);
+  } // General case
   if (dimension == 1) return distributionCollection_[0].computeCDF(point);
```
Comment (josephmure) [inline lib/src/Uncertainty/Distribution/JointDistribution.cxx:579]: Same

File: lib/src/Uncertainty/Distribution/JointDistribution.cxx [hunk 24]
```diff
@@ -387,35 +430,35 @@ struct ComposedDistributionComputeSamplePolicy
...
+    Sample result(size, dimension);
+    const ComposedDistributionComputeSamplePolicy policy( kernelSample, result, distributionCollection_ );
+    TBBImplementation::ParallelFor( 0, size, policy );
+    result.setName(getName());
+    result.setDescription(getDescription());
+    return result;
+  } // General case
   // For 1D or independent components, we can only rely on possible parallel
   // implementation of getSample() methods of the marginal distributions
   if (dimension == 1) return distributionCollection_[0].getSample(size);
```
Comment (josephmure) [inline lib/src/Uncertainty/Distribution/JointDistribution.cxx:447]: Same

File: lib/src/Uncertainty/Distribution/JointDistribution.cxx [hunk 25]
```diff
@@ -448,95 +491,96 @@ Point JointDistribution::computeDDF(const Point & point) const
...
+    {
+      component[0] = point[i];
+      uPoint[i] = distributionCollection_[i].computeCDF(component);
+      sumLogPDF += distributionCollection_[i].computeLogPDF(component);
+    }
+    return kernel_.computeLogPDF(uPoint) + sumLogPDF;
+  } // General case
   // Special case for dimension 1, to boost performances
-  if (dimension == 1) return distributionCollection_[0].computeLogPDF(point);
+  if ((dimension == 1) && (kernel_.isCopula())) return distributionCollection_[0].computeLogPDF(point);
```
Comment (josephmure) [inline lib/src/Uncertainty/Distribution/JointDistribution.cxx:558]: Same

File: lib/src/Uncertainty/Distribution/JointDistribution.cxx (hunk 19 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/JointDistribution.cxx:90]: No. Semi-copula is another concept (see the book of Nelsen, which is full of semi-copulas), and pseudo-copula is ugly. I don't want to invent a copula-related concept in the hurry of a RC just to tell that any distribution taking values in [0,1]^d can work. BTW:
  * The dimension can be equal to 1, and it is precisely the case when taking 1D margins of such distributions in the non-copula case
  * The range doesn't have to be exactly the unit cube but has to be included into it for the maths to work. A kind of truncation in the dependence part of the distribution. I will change the test accordingly.

**Comment by regislebrun** [comment]
@josephmure A second look at getRealization() made you change your mind ;-)?

File: lib/src/Uncertainty/Distribution/JointDistribution.cxx (hunk 19 above)
Comment (josephmure) [reply lib/src/Uncertainty/Distribution/JointDistribution.cxx:90]: Well, we might need to change it later if we come up with a name we can agree on. I will attempt just one additional name suggestion: "unitCubeDistribution_" (it still works even if the support only needs to be included in the unit cube).

**Comment by josephmure** [comment]
> @josephmure A second look at getRealization() made you change your mind ;-)?

Exactly!

File: lib/src/Uncertainty/Distribution/openturns/UniformOrderStatistics.hxx [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+  void computeMean() const override;
+
+  /** Compute the covariance of the distribution */
+  void computeCovariance() const override;
+
+  /** Get the quantile of the distribution */
+  Scalar computeScalarQuantile(const Scalar prob, const Bool tail = false) const override;
+
+  /** Compute the numerical range of the distribution given the parameters values */
+  void computeRange() override;
```
Comment (josephmure) [inline lib/src/Uncertainty/Distribution/openturns/UniformOrderStatistics.hxx:151]: Is there a particular reason why these methods are private rather than protected?

File: python/src/JointDistribution_doc.i.in (hunk 9 above)
Comment (josephmure) [inline python/src/JointDistribution_doc.i.in:19]:
```suggestion
copula, then it represents the dependence structure of the joint
```
Same

File: python/src/JointDistribution_doc.i.in [hunk 27]
```diff
@@ -67,3 +78,27 @@ Parameters
...
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::JointDistribution::getKernel
+"Get the kernel of the distribution.
+
+Returns
+-------
+kernel : :class:`~openturns.Distribution`
+    Kernel of the distribution. A distribution whose range is
+    included into the unit hypercube :math:`[0, 1]^d` where :math:`d`
```
Comment (josephmure) [inline python/src/JointDistribution_doc.i.in:91]:
```suggestion
    included into the unit hypercube :math:`[0, 1]^\inputDim` where :math:`\inputDim`
```

File: python/src/JointDistribution_doc.i.in [hunk 28]
```diff
@@ -67,3 +78,27 @@ Parameters
...
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::JointDistribution::setKernel
+"Set the kernel of the distribution.
+
+Parameters
+----------
+kernel : :class:`~openturns.Distribution`
+    Kernel of the distribution. Any distribution whose range is
+    included into the unit hypercube :math:`[0, 1]^d` where :math:`d`
```
Comment (josephmure) [inline python/src/JointDistribution_doc.i.in:103]:
```suggestion
    included into the unit hypercube :math:`[0, 1]^\inputDim` where :math:`\inputDim`
```

File: python/src/UniformOrderStatistics_doc.i.in [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+as (see [arnold2008]_ eq. 2.2.4 page 11):
+
+.. math::
+
+    f_{X_{(1:n)}}(x) = n!\,\mathbf{1}_{\cS}(x)
+
+where :math:`\cS\subset\Rset^n` is the simplex with vertices :math:`\cV`:
+
+.. math::
+   \cV=\left\{(\underbrace{0,\dots,0}_{k}, \underbrace{1,\dots,1}_{n-k}), k\in\{0,\dots,n\}\right\}`.
```
Comment (josephmure) [inline python/src/UniformOrderStatistics_doc.i.in:32]:
```suggestion
   \cV=\left\{(\underbrace{0,\dots,0}_{k}, \underbrace{1,\dots,1}_{n-k}), \; k\in\{0,\dots,n\}\right\}`.
```
Another unimplemented suggestion by @mbaudin47

File: python/src/UniformOrderStatistics_doc.i.in [hunk 30]
```diff
@@ -0,0 +1,54 @@
...
+
+The link between :math:`\cS` and the order statistics is given by:
+
+.. math::
+   \cS=\left\{(x_1,\dots,x_n)\in[0,1]^n\,|\,0\leq x_1\leq\dots\leq x_n\right\}=\left\{\sum_{k=1}^{n+1}\lambda_k v_k\,\left|\right.\, v_k\in\cV, \lambda_k\in[0,1], \sum_{k=1}^{n+1}\lambda_k=1\right\}
```
Comment (josephmure) [inline python/src/UniformOrderStatistics_doc.i.in:37]:
```suggestion
   \cS=\left\{(x_1,\dots,x_n)\in[0,1]^n\,|\,0 \leq x_1 \leq \dots \leq x_n \leq 1\right\}
   = \left\{\sum_{k = 1}^{n + 1} \lambda_k v_k \, \left|\right.\, v_k \in \cV, \; \lambda_k \in [0,1], \; \sum_{k = 1}^{n + 1} \lambda_k = 1\right\}
```
Same

File: python/src/UniformOrderStatistics_doc.i.in [hunk 31]
```diff
@@ -0,0 +1,54 @@
...
+>>> import openturns.experimental as otexp
+>>> distribution = otexp.UniformOrderStatistics(3)
+
+Draw a sample:
+
+>>> sample = distribution.getSample(5)
+
+Compute some quantities:
+
+>>> pdf = distribution.computePDF([0.1, 0.15, 0.45])
```
Comment (josephmure) [inline python/src/UniformOrderStatistics_doc.i.in:53]:
```suggestion
>>> pdf = distribution.computePDF([0.1, 0.15, 0.45])  // Notice the order
```
Same

**Comment by regislebrun** [comment]
@josephmure @mbaudin47 Sorry for the suggestions I dropped, I found several critical bugs this WE and I spent too much time trying to solve them. At the end of the day my eyes were no more in place and I missed things... Thanks for the review!

File: python/src/UniformOrderStatistics_doc.i.in [hunk 32]
```diff
@@ -0,0 +1,55 @@
...
+>>> import openturns.experimental as otexp
+>>> distribution = otexp.UniformOrderStatistics(3)
+
+Draw a sample:
+
+>>> sample = distribution.getSample(5)
+
+Compute some quantities:
+
+>>> pdf = distribution.computePDF([0.1, 0.15, 0.45])  // Notice the order
```
Comment (jschueller) [inline python/src/UniformOrderStatistics_doc.i.in:54]: thats the c++ comment syntax :]

File: lib/src/Uncertainty/Distribution/Beta.cxx [hunk 33]
```diff
@@ -289,27 +289,27 @@ void Beta::computeMean() const
...
 {
-  return Point(1, 2.0 * ((alpha_ + beta_) - 2.0 * alpha_) / ((alpha_ + beta_) + 2.0) * std::sqrt(((alpha_ + beta_) + 1.0) / (alpha_ * (beta_))));
+  return Point(1, 2.0 * ((alpha_ + beta_) - 2.0 * alpha_) / ((alpha_ + beta_) + 2.0) * std::sqrt(((alpha_ + beta_) + 1.0) / (alpha_ * beta_)));
 }
 
 /* Get the kurtosis of the distribution */
 Point Beta::getKurtosis() const
 {
-  return Point(1, 3.0 * (1.0 + (alpha_ + beta_)) * (2.0 * (alpha_ + beta_) * (alpha_ + beta_) + alpha_ * ((alpha_ + beta_) - 6.0) * (beta_)) / (alpha_ * (beta_) * (3.0 + (alpha_ + beta_)) * (2.0 + (alpha_ + beta_))));
+  return Point(1, 3.0 * (1.0 + (alpha_ + beta_)) * (2.0 * (alpha_ + beta_) * (alpha_ + beta_) + alpha_ * ((alpha_ + beta_) - 6.0) * beta_) / (alpha_ * beta_ * (3.0 + (alpha_ + beta_)) * (2.0 + (alpha_ + beta_))));
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Distribution/Beta.cxx:304]: There are still extra parentheses around the sum alpha + beta.

**Review by mbaudin47 — CHANGES_REQUESTED** [review]

**Review by mbaudin47 — CHANGES_REQUESTED** [review]

File: lib/src/Uncertainty/Distribution/Dirichlet.cxx [hunk 34]
```diff
@@ -469,7 +470,7 @@ Point Dirichlet::getSkewness() const
   for (UnsignedInteger i = 0; i < dimension; ++i)
   {
     const Scalar thetaI = theta_[i];
-    skewness[i] = 2.0 * (sumTheta_ - 2.0 * thetaI) / (sumTheta_ + 2.0) * std::sqrt(sumTheta_ + 1.0) / (thetaI * (sumTheta_ - thetaI));
+    skewness[i] = 2.0 * (sumTheta_ - 2.0 * thetaI) / (sumTheta_ + 2.0) * std::sqrt((sumTheta_ + 1.0) / (thetaI * (sumTheta_ - thetaI)));
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Distribution/Dirichlet.cxx:473]: This is not cosmetic anymore. Is this on purpose?

**Review by mbaudin47 — CHANGES_REQUESTED** [review]

**Review by josephmure — APPROVED** [review]
I still dislike the name "core" but it is not a blocking issue.

File: lib/src/Uncertainty/Distribution/Dirichlet.cxx (hunk 34 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Distribution/Dirichlet.cxx:473]: It is on purpose and fixes a bug.
